### PR TITLE
Fix unassigned ml system shard replicas

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description
In previous ml system shard replicas always stays on an arbitrary number and cannot expand automatically, which will cause unassigned shards and yellow cluster problem. This PR can resolve this issue.
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 1]
```diff
@@ -68,6 +68,7 @@ public void initMLConfigIndex(ActionListener<Boolean> listener) {
     public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener) {
         String indexName = index.getIndexName();
         String mapping = index.getMapping();
+        final Map<String, Object> indexSettings = Map.of("index.auto_expand_replicas", "0-6");
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:71]: So the maximum replica is 6?

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 2]
```diff
@@ -97,12 +98,19 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
...
+                                        updateSettingRequest.indices(indexName).settings(indexSettings);
+                                        client
+                                            .admin()
+                                            .indices()
+                                            .updateSettings(updateSettingRequest, ActionListener.wrap(updateResponse -> {
+                                                if (response.isAcknowledged()) {
+                                                    indexMappingUpdated.get(indexName).set(true);
+                                                    internalListener.onResponse(true);
+                                                } else {
+                                                    internalListener.onFailure(new MLException("Failed to update index: " + indexName));
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:111]: Tune the error message 
```suggestion
                                                    internalListener.onFailure(new MLException("Failed to update index setting: " + indexName));
```

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 3]
```diff
@@ -97,12 +98,19 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
...
+                                                }
+                                            }, internalListener::onFailure));
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:113]: Suggest log the error first, check line 115

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 1 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java]: Seems no need to recreate this map for every request, create a constant field?

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 4]
```diff
@@ -68,6 +68,7 @@ public void initMLConfigIndex(ActionListener<Boolean> listener) {
     public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener) {
         String indexName = index.getIndexName();
         String mapping = index.getMapping();
+        final Map<String, Object> indexSettings = Map.of("index.auto_expand_replicas", "0-5");
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:71]: Can we add any comment why we chose `0-5`?

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 5]
```diff
@@ -97,12 +98,19 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
...
+                                        updateSettingRequest.indices(indexName).settings(indexSettings);
+                                        client
+                                            .admin()
+                                            .indices()
+                                            .updateSettings(updateSettingRequest, ActionListener.wrap(updateResponse -> {
+                                                if (response.isAcknowledged()) {
+                                                    indexMappingUpdated.get(indexName).set(true);
+                                                    internalListener.onResponse(true);
+                                                } else {
+                                                    internalListener.onFailure(new MLException("Failed to update index setting: " + indexName));
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:111]: How about:

`internalListener.onFailure(new MLException("Failed to update index settings for: " + indexName));`

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 3 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:113]: +1

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 6]
```diff
@@ -41,8 +40,9 @@ public class MLIndicesHandler {
 
     private static final Map<String, AtomicBoolean> indexMappingUpdated = new HashMap<>();
     static {
-        indexMappingUpdated.put(ML_MODEL_INDEX, new AtomicBoolean(false));
-        indexMappingUpdated.put(ML_TASK_INDEX, new AtomicBoolean(false));
+        for (MLIndex mlIndex : MLIndex.values()) {
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:44]: This is suggested by @ylwu-amzn

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 5 above)
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java]: Looks good, I will use this

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 4 above)
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java]: To be consistent with AOS default settings

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:44]: MLIndex is enum. We need to get all index names and add to `indexMappingUpdated`

**Review by ylwu-amzn — DISMISSED** [review]

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 7]
```diff
@@ -97,12 +98,23 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
                                 .putMapping(
                                     new PutMappingRequest().indices(indexName).source(mapping, XContentType.JSON),
                                     ActionListener.wrap(response -> {
-                                        if (response.isAcknowledged()) {
-                                            indexMappingUpdated.get(indexName).set(true);
-                                            internalListener.onResponse(true);
-                                        } else {
-                                            internalListener.onFailure(new MLException("Failed to update index: " + indexName));
-                                        }
+                                        UpdateSettingsRequest updateSettingRequest = new UpdateSettingsRequest();
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:101]: I see you removed `if (response.isAcknowledged()) {`, I think we should keep this checking

Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:101]: In line 107

Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:101]: Or you think we need to check twice?

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:44]: Oh I see. That's what I was curious about.

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java (hunk 7 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:101]: If index mapping not sucessfull, no need to update setting, I think we should twice

**Review by dhrubo-os — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java [hunk 8]
```diff
@@ -98,10 +99,23 @@ public void initMLIndexIfAbsent(MLIndex index, ActionListener<Boolean> listener)
                                     new PutMappingRequest().indices(indexName).source(mapping, XContentType.JSON),
                                     ActionListener.wrap(response -> {
                                         if (response.isAcknowledged()) {
-                                            indexMappingUpdated.get(indexName).set(true);
-                                            internalListener.onResponse(true);
-                                        } else {
-                                            internalListener.onFailure(new MLException("Failed to update index: " + indexName));
+                                            UpdateSettingsRequest updateSettingRequest = new UpdateSettingsRequest();
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java:102]: For for this line, but you need to also add back these
```
else {	                                            client
                                            internalListener.onFailure(new MLException("Failed to update index: " + indexName));
```

**Review by ylwu-amzn — APPROVED** [review]
LGTM. Thanks for the change!

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**